python/hail/type.py

```diff
@@ -1,12 +1,582 @@
+            return TString()
+        elif class_name == 'is.hail.expr.TBoolean$':
+            return TBoolean()
+        elif class_name == 'is.hail.expr.TSample$':
+            return TString()
+        elif class_name == 'is.hail.expr.TArray':
+            return TArray._from_java(jtype)
+        elif class_name == 'is.hail.expr.TSet':
+            return TSet._from_java(jtype)
+        elif class_name == 'is.hail.expr.TDict':
+            return TDict._from_java(jtype)
+        elif class_name == 'is.hail.expr.TStruct':
+            return TStruct._from_java(jtype)
+        elif class_name == 'is.hail.expr.TVariant$':
+            return TVariant()
+        elif class_name == 'is.hail.expr.TAltAllele$':
+            return TAltAllele()
+        elif class_name == 'is.hail.expr.TLocus$':
+            return TLocus()
+        elif class_name == 'is.hail.expr.TInterval$':
+            return TInterval()
+        elif class_name == 'is.hail.expr.TGenotype$':
+            return TGenotype()
+        else:
+            raise TypeError("unknown type class: '%s'" % class_name)
+
+    def _jtype(self):
+        """
+        Returns the java type associated with this type.
+        """
+
+        raise NotImplemented('use a subclass of Type')
+
+    def _typecheck(self, annotation):
+        """
+        Raise an exception if the given annotation is not the appropriate type.
+
+        :param annotation: value to check
+        """
+        raise NotImplemented('use a subclass of Type')
+
+
+class TInt(Type):
+    """
+    Hail type corresponding to int
+    """
+
+    def __init__(self):
+        super(TInt, self).__init__(scala_object(Env.hail_package().expr, 'TInt'))
+
+    def _convert_to_py(self, annotation):
+        return annotation
+
+    def _convert_to_j(self, annotation):
+        if annotation:
+            return scala_package_object(Env.hail_package().utils).makeInt(annotation)
+        else:
+            return annotation
+
+    def _typecheck(self, annotation):
+        if annotation and not isinstance(annotation, int):
+            raise TypeCheckError("TInt expected type 'int', but found type '%s'" % type(annotation))
+
+
+class TLong(Type):
+    """
+    Hail type corresponding to long
+    """
+
+    def __init__(self):
+        super(TLong, self).__init__(scala_object(Env.hail_package().expr, 'TLong'))
+
+    def _convert_to_py(self, annotation):
+        return annotation
+
+    def _convert_to_j(self, annotation):
+        if annotation:
+            return scala_package_object(Env.hail_package().utils).makeLong(annotation)
+        else:
+            return annotation
+
+    def _typecheck(self, annotation):
+        if annotation and not (isinstance(annotation, long) or isinstance(annotation, int)):
+            raise TypeCheckError("TLong expected type 'int' or 'long', but found type '%s'" % type(annotation))
+
+
+class TFloat(Type):
+    """
+    Hail type corresponding to float
```

Comment:
Again, I don't think the size of Python floats is specified.  TFloat is 32-bit floating-point number.